File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/DiffCallback.java
package com.tunjid.androidbootstrap.recyclerview;

import com.tunjid.androidbootstrap.functions.BiFunction;
import com.tunjid.androidbootstrap.functions.Function;

import java.util.List;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

class DiffCallback<T> extends DiffUtil.Callback {

    private final List<T> stale;
    private final List<T> updated;
    private final Function<T, Differentiable> diffFunction;

    DiffCallback(List<T> stale, List<T> updated, Function<T, Differentiable> diffFunction) {
        this.stale = stale;
        this.updated = updated;
        this.diffFunction = diffFunction;
    }

    @Override
    public int getOldListSize() {
        return stale.size();
    }

    @Override
    public int getNewListSize() {
        return updated.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return apply(oldItemPosition, newItemPosition, (stale, current) -> stale.getId().equals(current.getId()));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return apply(oldItemPosition, newItemPosition, Differentiable::areContentsTheSame);
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        return apply(oldItemPosition, newItemPosition, Differentiable::getChangePayload);
    }

    private <S> S apply(int oldPosition, int newPosition, BiFunction<Differentiable, Differentiable, S> function) {
        return function.apply(diffFunction.apply(stale.get(oldPosition)), diffFunction.apply(updated.get(newPosition)));
    }
}

File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/diff/DiffCallback.kt
package com.tunjid.androidbootstrap.recyclerview.diff

import androidx.recyclerview.widget.DiffUtil

internal class DiffCallback<T>(
        private val stale: List<T>,
        private val updated: List<T>,
        private val diffFunction: (T) -> Differentiable
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = stale.size

    override fun getNewListSize(): Int = updated.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
            using(oldItemPosition, newItemPosition) { stale, current -> stale.diffId == current.diffId }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
            using(oldItemPosition, newItemPosition, Differentiable::areContentsTheSame)

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? =
            using(oldItemPosition, newItemPosition, Differentiable::getChangePayload)

    private fun <S> using(oldPosition: Int, newPosition: Int, function: (Differentiable, Differentiable) -> S): S =
            function(diffFunction(stale[oldPosition]), diffFunction(updated[newPosition]))
}
